### ISSUE TITLE
Feature Request: Presentation Mode

### ISSUE THREAD

**Issue by udfalkso** [yhatt/marp#13]
It would be great if I could present my presentation without the markdown showing and without having to export this to a file.  Much like Powerpoint's presentation mode.  
- Go full screen
- Support up/down left/right arrows and spacebar for navigating slides.  
- Esc to quit and return to editing mode.

**Comment by fabiantheblind** [comment]
plus speaker notes

**Comment by nyaray** [comment]
@fabiantheblind in the name of keeping things simple, I'd say that speaker notes probably is a feature of it's own (a new issue, perhaps?).

They require a new window (speaker view) to appear and having that synced with the main window (presentation view), so I'd say there's a bit of hidden work in that feature :P.

**Comment by fabiantheblind** [comment]
@nyaray done

**Comment by yhatt** [comment]
I'm sure anyone think this issue. So am I. (yhatt/mdslide#5)

I also had thought the other day that delivering a presentation just with Marp might be super-useful.
A my co-worker made a presentation (of course created by Marp) yesterday, but he showed the screen of 1:1 slide view mode to the audience while switching slide by moving editor's caret. :sweat:

Currently we are focusing enhancements as the presentation writer. But its maturation would be a trigger of implementing the presentation mode!

**Comment by nyaray** [comment]
@yhatt would you mind merging a PR for this? I'd really like to use this for presentations and am willing to figure out what needs to be done if you are too short on time to explain.

**Comment by yhatt** [comment]
@nyaray OK, I'll merge. I made clear milestones. We would aim it to the first minor release.

**Comment by will-hart** [comment]
@nyaray have you had a chance to look at this? I've been trying to work with exported PDFs but the exported image quality is very low. 

I thought as a concept something like the following:

- User selects "presentation mode" in the main window (win1) 
- win1 sends an IPC to the browser process including the currently selected slide contents
- browser process opens win2
- browser process sends through the slide contents via IPC to win2
- win 2 opens as a 1:1 slide view which renders the current slide contents passed by IPC
- win1 displays presentation mode, which includes speaker notes, navigation controls and views of the current and next slide

- When the user clicks next/previous in win1 it sends an IPC to the browser process containing the HTML (or Markdown?) of the new slide to show
- win2 receives the forwarded IPC and updates the displayed content
- if the users clicks "end presentation" on win1, an IPC is sent and win2 is closed. Win1 comes out of presentation mode

A fair bit of work I think in hooking up the IPCs, but it could possibly use the existing code for 1:1 previews so would prevent some duplication. 

There could possibly be an issue with running multiple marp windows, so maybe would need to somehow address the IPC directly to a particular window.

**Comment by nyaray** [comment]
@will-hart Nope, no work quite yet; I've been caught up with work as I unexpectedly found myself in extra responsibilities. Also, marp's missing features are a bit of an excuse to not finalise a talk I'm planning on giving at work ...

I've been working with node.js at work, so I guess it'll be useful for contributing to this, especially since we've done some IPC stuff :). I agree, there's quite a bit of synchronisation to do! It would make sense to break it down into parts though, in order to not create a mess in the code.

I suggest breaking the work down into something like the following:

- [ ] Main window presentation mode - fullscreen(?), only showing content
- [ ] Change slides with the arrow keys - make sure that 
- [ ] Presenter window, initially just follows the main window
  - Stuff to do, which likely can be done in any order/at the same time:
  - [ ] Add note displaying, maybe extracted from the slide comments
  - [ ] Share control over slides, next/previous
  - [ ] Display next (add option for previous?) slide
  - [ ] HH:mm-timer with Play/pause and reset

Any thoughts?

**Comment by gcnew** [comment]
I just hacked in some sort of presentation mode. It's not polished by any means, nor is it friendly to the rest of the code. I'd be happy to improve on it and send a real PR (if @yhatt is interested and helps with suggestions :)). In the mean time, if anyone is interested, here is my fork/commit:

https://github.com/gcnew/marp/commit/0dc3ee49658958106f3bdf661f439a9914c97134

Features:
 - button for presentation mode
 - exit presentation mode with escape
 - navigate with arrows

**Comment by mindplay-dk** [comment]
I have my speaker notes in `<!-- comments -->` inside my file.

I'm quite happy with that. I don't think we need any additional syntax or features for this?

I plan on simply cursoring through the file to track my progress while speaking - since the app will track and display the slide under the cursor, all that's really missing for me, is the ability to detach the slide view from the main window and full-screen it.

Being able to see only my notes would be a step down - being able to see the text and my notes together is more useful, as often the words on the slides are also cues.

In my opinion, that would be better than cursor-navigation, since I can track my progress not just per slide, but per line in the file.

That's much simpler, and it's already how you're used to working (and practicing) with your slides while you're creating them - having identical flows while creating and presenting in my opinion is much better than having to (mentally) switch modes.

Editing and previewing with Marp already (IMHO) is the best possible flow of any presentation software ever - being able to make minor corrections (typos etc.) instantaneously *while* presenting would be super powerful! Why change anything at all? :-)

**Comment by majkinetor** [comment]
+1

**Comment by cekvenich** [comment]
Wait, so how do I present?

**Comment by d0ugal** [comment]
> Wait, so how do I present?

I exported to PDF and used https://pdfpc.github.io/. It worked great tbh, I was skeptical but I was able to get a really good setup this way. It was a little more work than I'd like but the result was good.

**Comment by cekvenich** [comment]
ftw, I used this lib via npm to generate webpages to paginate/present, via Pug.
Example: install mbake via https://www.staticgen.com/metabake
and then mbake -l
will emit example with a readme file.
Happy to answer questions on my solution.

**Comment by yhatt** [comment]
There is a christmas gift from [marp-team](https://github.com/marp-team)! Today we have released the first stable CLI interface of [Marp Next](https://github.com/marp-team/marp) at [marp-team/marp-cli](https://github.com/marp-team/marp-cli).

It can output HTML file for presentation now. Open converted HTML in your Chrome / Firefox, or use Marp CLI's `--preview` option.

[![](https://raw.githubusercontent.com/marp-team/marp-cli/master/docs/images/marp-cli.gif)](https://github.com/marp-team/marp-cli)

By hitting <kbd>f</kbd> in the converted HTML,  you can present deck in full-screen.

It has not the rich features (e.g. speaker notes), but we think it fills our [MVP](https://en.wikipedia.org/wiki/Minimum_viable_product). We are going to work continously to support presentation mode in [Marp Web](https://github.com/marp-team/marp-web), the web-based GUI interface of Marp Next.

**Comment by cekvenich** [comment]
I tested it. VERY NICE!

It renders a bit different than the GUI tool.

**Comment by bkaradzic** [comment]
It would be really nice if there was presentation mode in editor (nothing complicated, just to put preview full screen and ability to flip slides). This extra CLI tool seems a bit too much for simple use case.

**Comment by yhatt** [comment]
@bkaradzic Yeah, it is still our goal. I'm focusing to enhance new Marp ecosystem ([Marp Next](https://github.com/marp-team/marp)), and CLI tool is a minimum requirement to present the deck in the new ecosystem.

I'm aiming to be easy to develop editor and presentation mode by simple modulized libraries (e.g. [Marp Core](https://github.com/marp-team/marp-core) + [Marp React](https://github.com/marp-team/marp-react) / [Marp Vue](https://github.com/marp-team/marp-vue)). Our roadmap has a plan to develop presentation mode in a renewed Marp editor, the migrated interface from desktop app to web-based React + PWA app.

However, I have to take care whole of Marp Next ecosystem, so just lack resources to develop editor (The member of Marp team for developing Marp Next tools is still only me). A renewed Marp would be different from desktop app, so I expect appearing the community-driven editor that has presentation mode.

**Comment by bkaradzic** [comment]
Still don't get it why all this has to happen before presentation mode is possible? It seems existing Marp contains everything needed for simple presentation mode (since already can preview slide). Only thing would be to add handling left/right buttons when presentation part of window is in focus.

Anyhow, yesterday I was just using external .pdf viewer (SumatraPDF) that automatically reload .pdf on changes. So streamlining export to .pdf and opening external PDF viewer would be sufficient from my POV.

**Comment by yhatt** [comment]
Warn again: [we have already dropped maintenance of current Marp](https://github.com/yhatt/marp/issues/267) because of several security risks. Don't expect to any update for yhatt/marp.

**Comment by bkaradzic** [comment]
Old Marp have advantage of being one simple app. This install node.js to run some CLI to see presentation is not better model. If I wanted to run server I would pick Go language present package and do it myself, definitely don't want Marp CLI / node-nonsense.js stuff...

I had pleasure working on my presentation yesterday with Old Marp... I downloaded, unziped, and was making stuff in no-time. That's actually **advantage** of Marp. But if you think that making it more complicated is what this project needs, go ahead, it's your project. :)

**Comment by yhatt** [comment]
That is our goal as I said. A planned (and re-newed) [Marp Desktop](https://github.com/marp-team/marp#apps) would replace for yhatt/marp. The old Marp looks like a simple, but has the worst maintainability by developer: Outdated language, no tests, and security risks.

I just have not time to work for Marp Desktop yet, that's all.

**Comment by yhatt** [comment]
**UPDATE:** Marp CLI v0.9.x has shipped together with standalone binaries. There is no requirements like [Node](https://github.com/marp-team/marp-cli#npx) and [Docker](https://github.com/marp-team/marp-cli#docker), and [just use the downloaded binary from release page](https://github.com/marp-team/marp-cli/releases).

Below command would open a window with ready to presentation in Chrome:

```
marp -ps --no-output [your-slide-deck.md]
```

**Comment by harsh183** [comment]
Glad to see there is still ongoing development on this. What's the rewrite version's take on remote support?
